reference/filone/reference.py

```
@@ -0,0 +1,319 @@
+        raise RuntimeError(error.stderr.strip() or error.stdout.strip() or "cast keccak failed") from error
+    if not HASH.fullmatch(result):
+        raise RuntimeError("cast keccak returned an invalid bytes32 value")
+    return result
+
+
+def load_product(path):
+    product = json.loads(Path(path).read_text(encoding="utf-8"))
+    exact_keys(product, set(PRODUCT_POLICY) | PRODUCT_NUMBERS, "product")
+    for key, expected in PRODUCT_POLICY.items():
+        if product[key] != expected:
+            raise ValueError(f"product.{key} must equal {expected!r}")
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)